### PR TITLE
Enable aux stack allocations on application heap

### PR DESCRIPTION
**Pull request by loganek**
This is necessary for WASI threads as the aux stack should be managed by the application. See https://github.com/bytecodealliance/wasm-micro-runtime/issues/1790 for details.

### CODE REVIEW

File: core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c [hunk 1]
```diff
@@ -527,9 +527,6 @@ pthread_start_routine(void *arg)
     /* destroy pthread key values */
     call_key_destructor(exec_env);
 
-    /* routine exit, destroy instance */
-    wasm_runtime_deinstantiate_internal(module_inst, true);
-
```
Comment (yamt) [inline core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c:532]: how is this related to the rest of the PR?

File: core/iwasm/libraries/thread-mgr/thread_manager.c [hunk 2]
```diff
@@ -77,8 +77,18 @@ traverse_list(bh_list *l, list_visitor visitor, void *user_data)
...
 static bool
-allocate_aux_stack(WASMCluster *cluster, uint32 *start, uint32 *size)
+allocate_aux_stack(WASMExecEnv *exec_env, uint32 *start, uint32 *size)
 {
+    WASMCluster *cluster = wasm_exec_env_get_cluster(exec_env);
+#if WASM_ENABLE_HEAP_AUX_STACK_ALLOCATION == 1
+    WASMModuleInstanceCommon *module_inst =
+        wasm_exec_env_get_module_inst(exec_env);
+
+    *start = wasm_runtime_module_malloc(module_inst, cluster->stack_size, NULL);
```
Comment (yamt) [inline core/iwasm/libraries/thread-mgr/thread_manager.c:87]: i'm not sure what's the intention of this.
for wasi-threads, we will disable this allocation, right?

Comment (loganek) [reply core/iwasm/libraries/thread-mgr/thread_manager.c:87]: Yes, it will be disabled for wasi threads, but even with that, we need to keep `wasm_cluster_spawn_exec_env()` working; aux stack allocation is needed for that.

File: core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c (hunk 1 above)
Comment (loganek) [reply core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c:532]: If `WASM_ENABLE_HEAP_AUX_STACK_ALLOCATION` is enabled, module instance can't be deinstantiate untill the memory allocated for the aux stack is released.

Comment (yamt) [reply core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c:532]: * with this change, where will it be deinstantiated instead?
* how about the other wasm_runtime_deinstantiate_internal call in this function? (L507)

File: core/iwasm/libraries/thread-mgr/thread_manager.c (hunk 2 above)
Comment (yamt) [reply core/iwasm/libraries/thread-mgr/thread_manager.c:87]: ok

File: core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c (hunk 1 above)
Comment (loganek) [reply core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c:532]: > with this change, where will it be deinstantiated instead?
It's deinstantiated in the `thread_manager_start_routine()`

> how about the other wasm_runtime_deinstantiate_internal call in this function? (L507)
Yes, this one, and one in the `lib_wasi_threads_wrapper.c` have to be removed. Already updated the code, thanks.

Comment (yamt) [reply core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c:532]: do you mean already updated locally but not pushed yet?

Comment (loganek) [reply core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c:532]: Yes sorry, didn't push my change. Should be updated now.

File: core/iwasm/libraries/lib-wasi-threads/lib_wasi_threads.cmake [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 set (LIB_WASI_THREADS_DIR ${CMAKE_CURRENT_LIST_DIR})
 
-add_definitions (-DWASM_ENABLE_LIB_WASI_THREADS=1)
+add_definitions (-DWASM_ENABLE_LIB_WASI_THREADS=1 -DWASM_ENABLE_HEAP_AUX_STACK_ALLOCATION)
```
Comment (wenyongh) [inline core/iwasm/libraries/lib-wasi-threads/lib_wasi_threads.cmake:6]: Should be `-DWASM_ENABLE_HEAP_AUX_STACK_ALLOCATION=1`

File: core/iwasm/libraries/thread-mgr/thread_manager.c [hunk 4]
```diff
@@ -77,8 +77,18 @@ traverse_list(bh_list *l, list_visitor visitor, void *user_data)
 }
 
 static bool
-allocate_aux_stack(WASMCluster *cluster, uint32 *start, uint32 *size)
+allocate_aux_stack(WASMExecEnv *exec_env, uint32 *start, uint32 *size)
 {
+    WASMCluster *cluster = wasm_exec_env_get_cluster(exec_env);
+#if WASM_ENABLE_HEAP_AUX_STACK_ALLOCATION == 1
```
Comment (wenyongh) [inline core/iwasm/libraries/thread-mgr/thread_manager.c:83]: Had better `#if WASM_ENABLE_HEAP_AUX_STACK_ALLOCATION != 0`

File: core/iwasm/libraries/thread-mgr/thread_manager.c [hunk 5]
```diff
@@ -100,11 +110,21 @@ allocate_aux_stack(WASMCluster *cluster, uint32 *start, uint32 *size)
...
     os_mutex_unlock(&cluster->lock);
     return false;
+#endif
 }
 
 static bool
-free_aux_stack(WASMCluster *cluster, uint32 start)
+free_aux_stack(WASMExecEnv *exec_env, uint32 start)
 {
+#if WASM_ENABLE_HEAP_AUX_STACK_ALLOCATION == 1
```
Comment (wenyongh) [inline core/iwasm/libraries/thread-mgr/thread_manager.c:119]: Same as above
